src/protocols/output_management.rs

```diff
@@ -0,0 +1,785 @@
+                            heads: HashMap::new(),
+                        },
+                    );
+                } else {
+                    error!("CreateConfiguration: missing client data");
+                }
+            }
+            zwlr_output_manager_v1::Request::Stop => {
+                let clients = &mut state.output_management_state().clients;
+                clients.get(&client.id()).map(|c| c.manager.finished());
+                clients.remove(&client.id());
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<D> Dispatch<ZwlrOutputConfigurationV1, (), D> for OutputManagementManagerState
+where
+    D: Dispatch<ZwlrOutputConfigurationV1, ()>,
+    D: Dispatch<ZwlrOutputHeadV1, String>,
+    D: Dispatch<ZwlrOutputConfigurationHeadV1, ZwlrOutputConfigurationV1>,
+    D: OutputManagementHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        client: &Client,
+        conf: &ZwlrOutputConfigurationV1,
+        request: zwlr_output_configuration_v1::Request,
+        _data: &(),
+        _display: &DisplayHandle,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        let g_state = state.output_management_state();
+        let client_data = match g_state.clients.get_mut(&client.id()) {
+            Some(data) => data,
+            _ => {
+                error!("OutputConfiguration: missing client data");
+                return;
+            }
+        };
+        let conf_data = match client_data.confs.get_mut(&conf.id()) {
+            Some(data) => data,
+            _ => {
+                error!("OutputConfiguration: missing conf data");
+                return;
+            }
+        };
+        let outdated = conf_data.serial != g_state.serial;
+        match request {
+            zwlr_output_configuration_v1::Request::EnableHead { id, head } => {
+                let head_name = match head.data::<String>() {
+                    Some(name) => name,
+                    None => {
+                        error!("EnableHead: Missing attached output head name");
+                        let _fail = data_init.init(id, conf.clone());
+                        return;
+                    }
+                };
+                if outdated {
+                    info!("EnableHead: request from an outdated configuration");
+                    let _fail = data_init.init(id, conf.clone());
+                    return;
+                }
+                let output = match conf_data.new_out.get_mut(head_name) {
+                    Some(out) => out,
+                    None => {
+                        error!("EnableHead: output configuration missing requested new head",);
+                        let _fail = data_init.init(id, conf.clone());
+                        return;
+                    }
+                };
+                if output.current_mode.is_none() {
+                    output.current_mode = output.modes.iter().position(|a| a.is_preferred);
```

Comment:
Shouldn't this do a similar logic to how the TTY backend picks a mode?

Reply:
like the `pick_mode()` in `backend/tty.rs` ?

Reply:
Yeah. But I haven't looked too deep into this yet to say that it's the right thing to do here.